### ISSUE TITLE
Saved through dill pipeline alters order of columns in transformed dataframe

### ISSUE THREAD

**Issue by yolking** [online-ml/river#1260]
Hello, I stumbled upon specific issue. After estimating pipeline and saving it to file/loading model through `dill`/`joblib` I get different column order and SGDRegressor doesn't like that. 
It's difficult to replicate on simple demo, here I came up with demo which triggers this in around 30% tries. It seems it's also session specific, cause if I just loop same code it always results in same (right or wrong) order of columns:

```
from sklearn import datasets
from sklearn.linear_model import SGDRegressor
from river import preprocessing
from river.compose import Select
from river import compat, compose
import numpy as np
import pandas as pd
import joblib

X,y = datasets.make_regression(n_samples=5, n_features=12,random_state=50)
X = pd.DataFrame(X)
X.columns = ['1','2','3','4','5','6','7','8', '9',  '10', '11', '12']
y = pd.Series(y)

model3 = Select('1','2','3','4','5','6','7','8', '9', '10', '11', '12') | preprocessing.StandardScaler()
model = ( model3|
    compat.convert_sklearn_to_river(SGDRegressor(tol=1e-10))

)
model.predict_many(X)
model.learn_many(X,y)

col_names_orig = model.transform_many(X).columns
dill.settings['ignore'] = False
with open('test.sav', 'wb') as file:
    dill.dump(model, file)
with open('test.sav', 'rb') as file:
    model1 = dill.load(file)

col_names_loaded = model1.transform_many(X).columns
if not (col_names_loaded == col_names_orig).all():
    print(col_names_loaded)
    print(col_names_orig)
model1.predict_many(X)
```
what breaks
`model1.predict_many(X)`
```
ValueError: The feature names should match those that were passed during fit.
Feature names must be in the same order as they were in fit.
```

**Comment by yolking** [comment]
It seems issue is related to `set` inside `Select` transformer

pre-dill model on call
`model.steps['TransformerUnion'].transformers['Pipeline2'].steps['Select'].keys`
returns 
`{'Past_2', 'CR', 'PP', 'Ef', 'Bo', 'F', 'Past_1', 'CC', 'Vo', 'St'}`
And after-dill model returns 
`{'Past_2', 'CR', 'PP', 'Vo', 'Ef', 'F', 'Past_1', 'CC', 'Bo', 'St'}`

**Comment by MaxHalford** [comment]
Thanks for catching this bug.

The clean way to proceed is to make sure the underlying `SKL2RiverRegressor` aligns the features in the same manner, at each method call. This is now done in #1263.

**Comment by MaxHalford** [comment]
@yolking can I let you check if this fixes your issue?

**Comment by yolking** [comment]
Yes, thank you. 
It fixes my issue but it's not an ideal solution for my current case. I would like to point it out here maybe for someone else in the future, that when I have to update SGD model and check for convergence after `learn_many` call , I have to estimate MSE from `y` and model predictions and check difference compared to previous coefs. That means calling `model.predict_many()` once or multiple times until convergence is reached. 
But in my understanding calling `predict_many` multiple times will retrain transformers on same data and result in wrong `StandardScaler` transformation and maybe some others on whole dataset . That's why  `estimator.predict(model.transform_many(X))` call is better in this case for prediction only task. And that will require handling of column order by user if model is saved and loaded from file.
I don't see other reasons to bypass `predict_many()` so if this isn't issue for other models in `river` this solution should be fine.

**Comment by MaxHalford** [comment]
That's a common thing users want to do @yolking! Indeed, it makes sense to want to make a prediction that doesn't update the pipeline's state. For this, there is a [`pure_inference_mode` context manager](https://riverml.xyz/0.17.0/api/compose/pure-inference-mode/).

I checked, and this context manager didn't actually work when working with `*_many` methods. I just fixed that in #1264.
